utils/brush/proc_macros/internal.rs

```diff
@@ -0,0 +1,195 @@
+            }
+        });
+
+    // Move ink! attrs from internal trait to external
+    impl_item.items
+        .iter_mut()
+        .for_each(|mut item|
+            if let syn::ImplItem::Method(method) = &mut item {
+                let method_key = method.sig.ident.to_string();
+
+                if ink_methods.contains_key(&method_key) {
+                    // Internal attrs will override external, so user must include full declaration with ink(message) and etc.
+                    ink_methods.get_mut(&method_key).unwrap().attrs = extract_attr(&mut method.attrs, "doc");
+                    ink_methods.get_mut(&method_key).unwrap().attrs.append(&mut extract_attr(&mut method.attrs, "ink"));
+                }
+            });
+
+    let ink_methods_iter = ink_methods.iter().map(|(_, value)| value);
+
+    let self_ty = impl_item.self_ty.clone().as_ref().clone();
+    let draft_impl: ItemImpl = syn::parse2(quote! {
+        #(#impl_ink_attrs)*
+        impl #trait_ident for #self_ty {
+            #(#ink_methods_iter)*
+        }
+    }).unwrap();
+
+    // Evaluate selector and metadata_name for each method based on rules in ink!
+    let ink_impl = ::ink_lang_ir::ItemImpl::try_from(draft_impl).unwrap();
+    ink_impl.iter_messages().for_each(|message| {
+        let method = ink_methods.get_mut(&message.ident().to_string()).unwrap();
+        if message.user_provided_selector().is_none() {
+            let selector_u32 = u32::from_be_bytes(message.composed_selector().as_bytes().clone()) as usize;
+            let selector = format!("{:#010x}", selector_u32);
+
+            method.attrs.push(new_attribute(quote! {#[ink(selector = #selector)]}));
+        }
+        if message.metadata_name() == message.ident().to_string() {
+            let selector = format!("{}", message.metadata_name());
+
+            method.attrs.push(new_attribute(quote! {#[ink(metadata_name = #selector)]}));
+        }
+
+        let original_name = message.ident();
+        let inputs_params = message.inputs().map(|pat_type| &pat_type.pat);
+
+        method.default = Some(syn::parse2(quote! {
+            {
+                #trait_ident::#original_name(self #(, #inputs_params )* )
+            }
+        }).unwrap());
+    });
+
+    let ink_methods_iter = ink_methods.iter().map(|(_, value)| value);
+    let wrapper_trait_ident = format_ident!("{}_{}{}", BRUSH_PREFIX, trait_ident, WRAPPER_TRAIT_SUFFIX);
+    // We only want to use this implementation in case when ink-as-dependency for wrapper.
+    // It will provide methods with the same name like in initial trait.
+    let wrapper_impl: ItemImpl = syn::parse2(quote! {
+        #(#impl_ink_attrs)*
+        impl #wrapper_trait_ident for #self_ty {
+            #(#ink_methods_iter)*
+        }
+    }).unwrap();
+
+    let trait_name = ink_impl.trait_path()
+        .map(|path| path.segments.last().map(|seg| seg.ident.to_string()))
+        .flatten();
```

Comment:
is there some way to use `flat_map` or something like that? we're functional programmers here :D